### PR TITLE
Configuration improvements - Document types

### LINKED ISSUE #9330: [Core] Config Deployment Improvements
Deployments vs. local configs
See also https://github.com/pimcore/data-hub/issues/357#issuecomment-842412425

- [x] custom reports 
- [x] document types 
- [x] predefined properties
- [x] predefined asset meta-data
- [x] thumbnails
- [x] web2print config
- [x] static routes (no settings-store)
- [x] perspectives
- [x] custom views

### PR DESCRIPTION
**Pull request by weisswurstkanone**
partly resolves #9330

### CODE REVIEW

File: bundles/AdminBundle/Controller/Admin/Document/DocumentController.php [hunk 1]
```diff
@@ -686,6 +690,9 @@ public function docTypesAction(Request $request)
                 $data = $this->decodeJson($request->get('data'));
                 $id = $data['id'];
                 $type = Document\DocType::getById($id);
+                if (!$type->isWriteable()) {
```
Comment (dvesh3) [inline bundles/AdminBundle/Controller/Admin/Document/DocumentController.php:693]: so we allow modification or deletion from API irrespective of `iswritable()`?, like if i do:
```php
$type = Document\DocType::getById($id);
$type->delete()
```

Comment (weisswurstkanone) [reply bundles/AdminBundle/Controller/Admin/Document/DocumentController.php:693]: Currently yes. I would say we should discuss in a separate thread because we have to align this for other configurations as well.

File: models/Element/ConfigWriteException.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * - GNU General Public License version 3 (GPLv3)
+ * - Pimcore Commercial License (PCL)
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Model\Element;
```
Comment (brusch) [inline models/Element/ConfigWriteException.php]:
```suggestion
namespace Pimcore\Model\Exception;
```
Since this exception is not directly related to `Element` I'd put that into a more generic namespace.

File: models/Document/DocType/Dao.php [hunk 3]
```diff
@@ -56,6 +74,15 @@ public function getById($id = null)
...
     {
+        if (!$this->model->getId()) {
+            $listing = new Listing();
+            $listing = $listing->load();
+            $listing = array_map(function(Model\Document\DocType $item) {
+                return $item->getId();
+            }, $listing);
+            $id = $this->getNextId($listing);
+            $this->model->setId($id);
+        }
```
Comment (brusch) [inline models/Document/DocType/Dao.php]: Is this maybe something we can make generic in the `AutoIncrementTrait` by providing the listing class? 
```php
$this->getNextId(Listing::class);
```

File: bundles/CoreBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -728,6 +728,29 @@ private function addDocumentsNode(ArrayNodeDefinition $rootNode)
 
         $documentsNode
             ->children()
+                 ->arrayNode('doctype')
```
Comment (brusch) [inline bundles/CoreBundle/DependencyInjection/Configuration.php]:
```suggestion
                 ->arrayNode('doc_types')
```
Since we use camelCase for the class name, I'd use snake_case for the config node.

**Review by brusch — CHANGES_REQUESTED** [review]

File: doc/Development_Documentation/21_Deployment/03_Configuration_Environments.md [hunk 5]
```diff
@@ -35,7 +35,8 @@ The following options are available:
 
 Available environment variables are: 
 ```bash
-PIMCORE_WRITE_TARGET_IMAGE_THUMBNAILS=settings-store
 PIMCORE_WRITE_TARGET_CUSTOM_REPORTS=settings-store
+PIMCORE_WRITE_TARGET_DOCUMENT_TYPES=settings-store
```
Comment (jdreesen) [inline doc/Development_Documentation/21_Deployment/03_Configuration_Environments.md]: Can't add a suggestion, but this should be added in the list above (line 27), see: #10217

File: models/Document/DocType/Dao.php [hunk 6]
```diff
@@ -22,12 +22,22 @@
...
+    use Model\Dao\AutoIncrementTrait;
+
     public function configure()
     {
-        parent::configure();
-        $this->setFile('document-types');
+        $config = \Pimcore::getContainer()->getParameter("pimcore.config");
+
+        parent::configure([
+            'containerConfig' => $config['documents']['doc_types']['definitions'],
```
Comment (dvesh3) [inline models/Document/DocType/Dao.php]:
```suggestion
            'containerConfig' => $config['documents']['doc_types']['definitions'] ?? [],
```

File: bundles/AdminBundle/Resources/public/js/pimcore/settings/docTypes.js [hunk 7]
```diff
@@ -213,8 +214,14 @@ pimcore.settings.document.doctypes = Class.create({
                 menuText: t('delete'),
                 width: 30,
                 items: [{
+                    getClass: function (v, meta, rec) {
+                        var klass = "pimcore_action_column ";
+                        if (rec.data.writeable) {
+                            klass += "pimcore_icon_delete";
```
Comment (dvesh3) [inline bundles/AdminBundle/Resources/public/js/pimcore/settings/docTypes.js]: Use old icon class
```suggestion
                            klass += "pimcore_icon_minus";
```

**Review by dvesh3 — CHANGES_REQUESTED** [review]
On adding new doctype the row is greyed out and not editable:
image

note: environment var is not set

File: bundles/AdminBundle/Resources/public/js/pimcore/settings/docTypes.js [hunk 8]
```diff
@@ -269,12 +285,16 @@ pimcore.settings.document.doctypes = Class.create({
                     {
                         text: t('add'),
                         handler: this.onAdd.bind(this),
-                        iconCls: "pimcore_icon_add"
+                        iconCls: "pimcore_icon_add",
+                        disabled: !pimcore.settings['image-thumbnails-writeable']
```
Comment (brusch) [inline bundles/AdminBundle/Resources/public/js/pimcore/settings/docTypes.js]:
```suggestion
                        disabled: !pimcore.settings['document-types-writeable']
```

**Review by brusch — CHANGES_REQUESTED** [review]

File: bundles/AdminBundle/Controller/Admin/Document/DocumentController.php [hunk 9]
```diff
@@ -709,7 +724,10 @@ public function docTypesAction(Request $request)
 
                 $type->save();
 
-                return $this->adminJson(['data' => $type->getObjectVars(), 'success' => true]);
+                $responseData = $type->getObjectVars();
+                $responseData['writeable'] = $type->isWriteable();
+
+                return $this->adminJson(['data' => $responseData, 'success' => true]);
```
Comment (dvesh3) [inline bundles/AdminBundle/Controller/Admin/Document/DocumentController.php:733]: Return `writeable` property on update action as well
